### PR TITLE
let transformers hold more wires

### PR DESCRIPTION
**Pull request by lashtear**
This allows stuffing more wires into the transformer block for more complicated ratios and makes a (weak) stab at rendering the results.

screenshot

### CODE REVIEW

**Comment by AMIDIBOSS** [comment]
Great

**Comment by AMIDIBOSS** [comment]
I love it.

**Comment by AMIDIBOSS** [comment]
approved

File: src/main/java/mods/eln/transparentnode/transformer/TransformerContainer.java [hunk 1]
```diff
@@ -21,10 +21,10 @@
 
     public TransformerContainer(EntityPlayer player, IInventory inventory) {
         super(player, inventory, new Slot[]{
-            new SixNodeItemSlot(inventory, primaryCableSlotId, 58, 30, 4, new Class[]{ElectricalCableDescriptor.class},
+            new SixNodeItemSlot(inventory, primaryCableSlotId, 58, 30, 64, new Class[]{ElectricalCableDescriptor.class},
```
Comment (AMIDIBOSS) [inline src/main/java/mods/eln/transparentnode/transformer/TransformerContainer.java:24]: Isn't 64 like too much?

File: src/main/java/mods/eln/transparentnode/transformer/TransformerContainer.java [hunk 2]
```diff
@@ -21,10 +21,10 @@
...
                 SlotSkin.medium,
                 new String[]{tr("Electrical cable slot")}),
-            new SixNodeItemSlot(inventory, secondaryCableSlotId, 100, 30, 4,
+            new SixNodeItemSlot(inventory, secondaryCableSlotId, 100, 30, 64,
```
Comment (AMIDIBOSS) [inline src/main/java/mods/eln/transparentnode/transformer/TransformerContainer.java:27]: ?

File: src/main/java/mods/eln/transparentnode/transformer/TransformerContainer.java (hunk 1 above)
Comment (AMIDIBOSS) [inline src/main/java/mods/eln/transparentnode/transformer/TransformerContainer.java:24]: Does it "fit" on the core?

**Review by AMIDIBOSS — CHANGES_REQUESTED** [review]

**Comment by lashtear** [comment]
14 is the max for visually distinct rings.  15 they are starting to blend together, and at 16 it's a solid orange bar.  Higher than that and you start getting really ugly z-fighting.  I didn't worry about that too much as we could probably use a better model for the wire or draw it via a scaled block primitive (at least until we redo for 1.10+ models)

I can set the limit to be 14 we think that is a good idea; it has the nice side effect of being less than 16 and thus preventing people from skipping voltage tiers so easily using better transformer ratios...?

File: src/main/java/mods/eln/transparentnode/transformer/TransformerDescriptor.java [hunk 3]
```diff
@@ -83,22 +83,22 @@ void draw(Obj3DPart fero, int priCableNbr, int secCableNbr, boolean hasCasing, f
...
+                    GL11.glTranslatef (0f, separation, 0f);
                 }
                 GL11.glPopMatrix();
             }
             if (secCableNbr != 0) {
                 GL11.glPushMatrix();
                 GL11.glRotatef(180, 0f, 1f, 0f);
-                float y = (secCableNbr - 1) * 1f / 16f;
-                GL11.glTranslatef(0f, -y, 0f);
+                float separation = 0.5f / (secCableNbr + 1);
```
Comment (Grissess) [inline src/main/java/mods/eln/transparentnode/transformer/TransformerDescriptor.java:97]: A call to glScalef around here might accomplish two things:

- Thin out the models on the Y axis enough that they remain visually distinct at higher numbers;
- Get rid of this slightly complicated math to calculate a separation--the separation can be counted in the same units as the Y-width (height?) of the model itself :)

On that note, I wouldn't mind the stacking limit as being the limit for a transformer, at least for experimentation. To assuage concerns of tier-skipping, I'd rather introduce a config or make them undesirable at high ratios for some reason (e.g., they become lossy, they have high capacitance, etc.).

Comment (Grissess) [reply src/main/java/mods/eln/transparentnode/transformer/TransformerDescriptor.java:97]: (NB: I don't know where your transform origin is, and that *might* have an undesirable effect of needing to multiply various translate parameters in this method :)

Comment (lashtear) [reply src/main/java/mods/eln/transparentnode/transformer/TransformerDescriptor.java:97]: Already tried that; the existing wires (called sbire model in the code) are introduced as a regular object from blender, and pre-translated.  We'd have to un-translate first at and that point it's better to scrap it.

Comment (Grissess) [reply src/main/java/mods/eln/transparentnode/transformer/TransformerDescriptor.java:97]: I'm amiable to the "scrap it" route, it's definitely simple enough to be practical to draw right in the code :)

**Comment by AMIDIBOSS** [comment]
I thinking 16 is the best solution here.  Skiping steps shouldn't be a problem. Plus some transformer windings look like solid copper if you don't look close enough.

**Comment by jrddunbr** [comment]
I am considering merging this into LLv11 or LLv12. Any thoughts/changes desired?

**Comment by Baughn** [comment]
The rings are geometrically really simple. It should be possible to squash them proportionally to the number of rings, perhaps above a minimum threshold, using simple OpenGL scaling.

Something like this, maybe?

```
glScalef(1.0, 6.0 / Math.max(6.0, ringCount), 1.0)
```

Of course you'll need to preserve the matrix around that call, if that isn't already happening. And this is a rather inefficient way to use OpenGL. That shouldn't matter, unless people plop down hundreds of the things in a single chunk, but do optimize out the call for ringCounts below 7.

**Comment by Baughn** [comment]
We've got this in Erisia now. I'm still waiting for an update PR so I can merge into develop. :)

**Comment by jrddunbr** [comment]
This is in my branch, with updated rendering code that @Grissess wrote, so if/when you pull that in it will be good to go. He worked on the renderer a bit and made it technically support up to 64 cables, but we set the cap to 16.

**Comment by OmegaHaxors** [comment]
Would like to see it go to 64 to both increase the amount of ratios you can do, while also making it more expensive to use in the process. 16 is pretty acceptable right now though, and I've personally been using a lot of 15:16 transformers for small stepups.

**Comment by lashtear** [comment]
You need a rebase and... scaling?  That'll require un-translation, scaling, and re-translation of the ring model, unless someone wants to make a new ring asset.  Or we make one in code (it's just a squashed cube, ugh...)

**Comment by jrddunbr** [comment]
We already made these changes work for LL. It's also in the Erisia build now.

**Comment by OmegaHaxors** [comment]
You think we could have those thick cables back for the smaller ratios? 1:4 looks really silly right now.

**Comment by lashtear** [comment]
If it's merged everywhere everyone wants, shall I close the PR?

**Comment by jrddunbr** [comment]
I'm indifferent. I suspect we're going to be closing a lot of 1.7 tickets
with 1.12+

**Comment by lashtear** [comment]
Closing as obsolete / undesired, can revisit later.

**Pull request closed without merging**